### PR TITLE
TypeScript: convert `fontProvider`

### LINKED ISSUE #12663: TypeScript: Convert font provider

### PR DESCRIPTION
**Pull request by barklund**
## Context

Converting the font provider to TS.

## Todo
- [x] Convert to TypeScript
- [x] Wait for constants to be merged in #12652  (currently cause lint and type errors, but @miina already converted this elsewhere so just waiting for her to merge that before fixing this one).
- [ ] Check tests?

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12663

### CODE REVIEW

File: packages/elements/src/types/data.ts
Comment (barklund) [inline packages/elements/src/types/data.ts:76]: Does this look right?

File: packages/fonts/package.json
Comment (barklund) [inline packages/fonts/package.json:50]: This one stumped me quite a bit! It just kept refusing to load the types and I couldn't understand why... 🤦🤦🤦

File: packages/story-editor/src/utils/cleanForSlug.ts
Comment (barklund) [inline packages/story-editor/src/utils/cleanForSlug.ts:59]: Hmm, I expected that `.pop()` on a `[string]` tuple would be typed as always returning a string, but it's still typed as `string | undefined`. Annoying :/

File: packages/story-editor/src/app/font/types.ts
Comment (miina) [inline packages/story-editor/src/app/font/types.ts]: There's also `FontStyle` in the `elements` package.

File: packages/story-editor/src/app/font/types.ts
Comment (miina) [inline packages/story-editor/src/app/font/types.ts]: Shouldn't these be `void` instead of `unknown`? No big difference I guess.
Also, now looking at it, `loadCustomFonts` and `loadCuratedFonts` or something similar would perhaps be more accurate names, `get*` indicates as if it would actually return something.

File: packages/story-editor/src/app/font/types.ts
Comment (miina) [inline packages/story-editor/src/app/font/types.ts:35]: In which case is it `null`? Saw that it's set to `[]` in the initial `state`, should it be `null` there instead perhaps?

File: packages/elements/src/types/data.ts
Comment (miina) [reply packages/elements/src/types/data.ts:76]: Seems right to me.

**Review by miina — COMMENTED** [review]
Looks great! Will check again and approve after merging `main` (although I suppose there won't be huge changes).

File: packages/story-editor/tsconfig.json
Comment (miina) [inline packages/story-editor/tsconfig.json]: These are covered by `"src/utils/*.ts",` above.

**Review by miina — APPROVED** [review]
Some nits, generally looks good.

File: packages/story-editor/src/utils/cleanForSlug.ts
Comment (spacedmonkey) [reply packages/story-editor/src/utils/cleanForSlug.ts:59]: I do that here - https://github.com/GoogleForCreators/web-stories-wp/pull/12737/files#diff-2c4c08b4833fb80bb4751a4252439074bd9c084802deafa659abd3a7d7cbcddfR45

File: packages/story-editor/tsconfig.json
Comment (barklund) [reply packages/story-editor/tsconfig.json]: Fixed in 9cffebe989,

File: packages/story-editor/src/app/font/types.ts
Comment (barklund) [reply packages/story-editor/src/app/font/types.ts:35]: Good point, fixed in 9cffebe989,

File: packages/story-editor/src/app/font/types.ts
Comment (barklund) [reply packages/story-editor/src/app/font/types.ts]: Good points, fixed in 9cffebe989,

File: packages/story-editor/src/app/font/types.ts
Comment (barklund) [reply packages/story-editor/src/app/font/types.ts]: Thanks, fixed in 9cffebe989,

File: packages/story-editor/src/app/font/types.ts
Comment (barklund) [reply packages/story-editor/src/app/font/types.ts:35]: It actually had to be `null`, otherwise we can't tell "custom fonts not loaded" from "no custom fonts". Now they're respectively `null` and empty list (again).

**Pull request merged**
